### DISCUSSION TITLE
AutoDiscoverFailed:

### DISCUSSION THREAD

**Discussion by bhaskarvenna** [ecederstrand/exchangelib#977] [Q&A]
Hi,
I was able to make connection and import messages earlier, and recently i had changed password , Now I'm getting below error for some reason AutoDiscoverFailed(pasted full message below). However when I manually login to my https://outlook.office365.com/mail/inbox, this is works fine.

Header:
UnauthorizedError: Invalid credentials for https://autodiscover-s.outlook.com/Autodiscover/Autodiscover.xml

Footer:
"_AutoDiscoverFailed: All steps in the autodiscover protocol failed for email 'diannaraedle@deerislegroup.com'. If you think this is an error, consider doing an official test at https://testconnectivity.microsoft.com_"

Can you please suggest me, how to resolve this issue.

Regards:
Bhasker.V
Jsglp

**Comment by ecederstrand** [comment]
What does the test at https://testconnectivity.microsoft.com say?

If you know that your account is hosted on Office365, there's not really any point in using autodiscover. Just connect to `outlook.office365.com` directly.

**Comment by bhaskarvenna** [comment]
Hi,
based on your comments i went ahead and did following tests and everything is passed except IMap where one of the service is failed.
https://testconnectivity.microsoft.com 
 
Also I'm not sure if I fully understood when you said connect, outlook.office364.com directly. Is there anything that i should change in python code to direct connection? 

_TEST RESULTS:_

Under Office 365, I was able to conduct below test
1.Exchange ActiveSync (ALL TEST PASSED)
2.Outlook Connectivity (ALL TEST PASSED)
3.POP Email (ALL TEST PASSED
4.Inbound SMTP Email (ALL TEST PASSED)
5.Synchronization, Notification, Availability, and Automatic Replies (This has Autodiscover as well) and ALL TEST PASSED
6.Inbound SMTP Email (ALL TEST PASSED)
7.Mailbox Provisioning Test (ALL TEST PASSED)
8.IMAP Email (ALL TEST PASSED , EXCEPT The IMAP service FAILED)

To my knowledge we haven't changed anything between when it was working fine before vs right now , except password.

Regards:
Bhasker.V
Jsglp

**Reply by bhaskarvenna** [reply]
In addition to previous now i'm getting following error.

_"UnauthorizedError: Invalid credentials for https://outlook.office365.com/EWS/Exchange.asmx"_

**Reply by ecederstrand** [reply]
If you already know the destination server, there's no need to go through autodiscover. Skipping autodiscovery is described in https://ecederstrand.github.io/exchangelib/#optimizing-connections

Are you running the latest version of exchangelib? if not, please update to the latest version and retry. I've seen this problem on older versions.

If you need autodiscovery and still have issues, your best bet is to compare the steps taken in the Autodiscover connectivity test report with the steps taken in exchangelib (you can see those by enabling debug logging). It's possible, for example, that you have TLS certificate validation issues locally due to missing root certificates on your local machine.
